### PR TITLE
WIP: A prototype for pod lifecycle event generator

### PR DESCRIPTION
**Pull request by yujuhong**
This is a prototype implementation for proposal #12802. Do not merge.

### CODE REVIEW

File: pkg/kubelet/dockertools/manager.go [hunk 1]
```diff
@@ -1676,3 +1676,93 @@ func getUidFromUser(id string) string {
 	// no gid, just return the id
 	return id
 }
+
+// A helper function to Get the pod/container information from the name of the
+// container.
+func GetPodInfoFromDockerName(rawName string) (*kubecontainer.Pod, string, uint64, error) {
```
Comment (vishh) [inline pkg/kubelet/dockertools/manager.go]: nit: `GetPodInfoFromDockerContainerName`

File: pkg/kubelet/dockertools/manager.go [hunk 2]
```diff
@@ -1676,3 +1676,93 @@ func getUidFromUser(id string) string {
...
+	Pod              *kubecontainer.Pod
+	IP               string
+	IsInfraContainer bool
+}
+
+// TODO(yujuhong): Refactor this function to share code with or replace
+// inspectContainer(). Note that this function doesn't handle tPath.
+func (dm *DockerManager) ExamineContainer(dockerID string) (*ContainerExaminationResult, error) {
+	result := &ContainerExaminationResult{}
+	iResult, err := dm.client.InspectContainer(dockerID)
```
Comment (vishh) [inline pkg/kubelet/dockertools/manager.go]: i like these `i` variable names :)

File: pkg/kubelet/pleg/docker.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+	containerSet util.StringSet
+}
+
+var _ PodLifecycleEventGenerator = &DockerPLEG{}
+
+func NewDockerPLEG(upstream ContainerEventWatcher, runtime *dockertools.DockerManager, channelCapacity int,
+	relistPeriod time.Duration) *DockerPLEG {
+	return &DockerPLEG{
+		relistPeriod: relistPeriod,
+		upstream:     upstream,
```
Comment (vishh) [inline pkg/kubelet/pleg/docker.go]: Is it possible to rename `upstream` to something else?

File: pkg/kubelet/pleg/docker.go [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+	// those events until relist is completed so that we can deduplicate
+	// events.
+
+	// Start watching upstream for container changes.
+	// TODO(yujuhong): Make sure there is enough channel capacity to buffer
+	// the events, or we should buffer it internally.
+	ch, err := d.upstream.Watch()
+	if err != nil {
+		// TODO(yujuhong): Should we retry?
+		glog.Errorf("Unable to watch upstream %v: %v", d.upstream, err)
```
Comment (vishh) [inline pkg/kubelet/pleg/docker.go]: May be retry after a delay?

File: pkg/kubelet/pleg/docker.go [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+	d.bootstrap()
+	glog.V(3).Infof("DockerPLEG: Bootstrapping completed.")
+	// Start a goroutine to process events.
+	go func() {
+		for {
+			select {
+			case e, ok := <-ch:
+				if !ok {
+					glog.Infof("DockerPLEG: Upstream channel closed")
+					return
```
Comment (vishh) [inline pkg/kubelet/pleg/docker.go]: Is it the responsibility of callers to handle event watch failures?

**Comment by vishh** [comment]
Did an initial cursory review. Will look into it in more detail once the WIP tag is removed.

**Comment by yujuhong** [comment]
> Did an initial cursory review. Will look into it in more detail once the WIP tag is removed.

Thanks for taking a look! I appreciate the feedback.

 This is still quite preliminary because things that need to be performed periodically would need to be handled properly -- either moving out as a separate component or enabled in the control flow. One such example is probing. We can't merge this PR without them properly handled.

**Comment by dchen1107** [comment]
I am going to close this pr since we carried it on with different prs, like #13571

**Pull request closed without merging**